pkg/postage/service.go

```diff
@@ -165,24 +153,36 @@ func (ps *service) IssuerUsable(st *StampIssuer) bool {
 func (ps *service) GetStampIssuer(batchID []byte) (*StampIssuer, func() error, error) {
 	ps.lock.Lock()
 	defer ps.lock.Unlock()
-	for _, st := range ps.issuers {
-		if bytes.Equal(batchID, st.data.BatchID) {
-			if !ps.IssuerUsable(st) {
-				return nil, nil, ErrNotUsable
-			}
-			return st, func() error {
-				return ps.save(st)
-			}, nil
-		}
+
+	if _, ok := ps.issuersInUse[string(batchID)]; ok {
+		return nil, nil, ErrBatchInUse
+	}
+	item := NewStampIssuerItem(batchID)
+	err := ps.store.Get(item)
+	if errors.Is(err, storage.ErrNotFound) {
+		return nil, nil, ErrNotFound
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if !ps.IssuerUsable(item.Issuer) {
+		return nil, nil, ErrNotUsable
 	}
-	return nil, nil, ErrNotFound
+	ps.issuersInUse[string(batchID)] = struct{}{}
+	return item.Issuer, func() error {
+		ps.lock.Lock()
+		defer ps.lock.Unlock()
+		delete(ps.issuersInUse, string(batchID))
+		return ps.save(item.Issuer)
+	}, nil
 }
 
-// save persists the specified stamp issuer to the statestore.
+// save persists the specified stamp issuer to the stamperstore.
 func (ps *service) save(st *StampIssuer) error {
-	st.bucketMu.Lock()
-	defer st.bucketMu.Unlock()
-	if err := ps.store.Put(ps.keyForIndex(st.data.BatchID), st); err != nil {
+	if err := ps.store.Put(&StampIssuerItem{
```

Comment:
can multiple parallel uploads using the same stamp issuer cause a data race when saving the issuer?

Reply:
That's what `issuersInUse` map solves. It prevents multiple routines from using the same issuer.

Reply:
@aloknerurkar is this ok?

Reply:
@istae So the problem is right now we have the transactions in the storer. But when we use the stampIssuer the data is in a different storage layer. As a result its very difficult to maintain consistency. Even if we don't call the Save function after using the issuer, the indexes that were assigned in-memory cannot be recovered if the transaction or the upload fails for eg. Also, on top of that, if we share it across upload sessions, it is impossible to recover indexes.

So we thought of keeping it simple and only allowing one upload per stamp at a time. I think the risk to reward is not that great here as if we have indexes that are not being used then the stamp is being wasted.

Reply:
Because we will no longer have in in-mem if the thing fails, the next upload will read the old state, so it wont lose anything.